Native executable cannot be built with signed Jars containing beans

### Describe the bug

Consider a Maven dependency which contains bean classes (e.g. `@ApplicationScoped`) and which is signed. Then use this dependency and one of its beans in a Quarkus project. Quarkus will create proxy classes for this bean at compile time and put them into `target/quarkus-app/quarkus/generated-bytecode.jar`. The package of these generated classes is the same as the package of the signed bean.
Now if you want to create a native executable this fails because the classloader during native compilation sees classes from the same package which are both signed (the original bean) and unsigned (the generated proxy classes). This result in a `SecurityException` and hence compilation fails.

### Expected behavior

The native executable can be compiled.

### Actual behavior

An error during native compilation such as
```
Error: Class initialization of io.quarkus.runner.ApplicationImpl failed. Use the option 

    '--initialize-at-run-time=io.quarkus.runner.ApplicationImpl'

 to explicitly request initialization of this class at run time.
com.oracle.svm.core.util.UserError$UserException: Class initialization of io.quarkus.runner.ApplicationImpl failed. Use the option 

    '--initialize-at-run-time=io.quarkus.runner.ApplicationImpl'

 to explicitly request initialization of this class at run time.
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.UserError.abort(UserError.java:85)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.classinitialization.ClassInitializationSupport.ensureClassInitialized(ClassInitializationSupport.java:195)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.classinitialization.AllowAllHostedUsagesClassInitializationSupport.computeInitKindAndMaybeInitializeClass(AllowAllHostedUsagesClassInitializationSupport.java:191)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.classinitialization.AllowAllHostedUsagesClassInitializationSupport.computeInitKindAndMaybeInitializeClass(AllowAllHostedUsagesClassInitializationSupport.java:129)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.classinitialization.ClassInitializationSupport.maybeInitializeAtBuildTime(ClassInitializationSupport.java:161)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.classinitialization.ClassInitializationSupport.maybeInitializeAtBuildTime(ClassInitializationSupport.java:150)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.SVMHost.onTypeReachable(SVMHost.java:310)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisUniverse.onTypeReachable(AnalysisUniverse.java:699)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisType.lambda$new$0(AnalysisType.java:310)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.AnalysisFuture.ensureDone(AnalysisFuture.java:63)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisType.ensureOnTypeReachableTaskDone(AnalysisType.java:696)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisType.onReachable(AnalysisType.java:590)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.AtomicUtils.atomicSetAndRun(AtomicUtils.java:49)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisType.lambda$registerAsReachable$8(AnalysisType.java:562)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisType.forAllSuperTypes(AnalysisType.java:676)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisType.forAllSuperTypes(AnalysisType.java:659)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisType.forAllSuperTypes(AnalysisType.java:655)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.meta.AnalysisType.registerAsReachable(AnalysisType.java:562)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.reflect.ReflectionDataBuilder.registerClass(ReflectionDataBuilder.java:204)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.reflect.ReflectionDataBuilder.lambda$register$2(ReflectionDataBuilder.java:173)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.executeCommand(CompletionExecutor.java:187)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.lambda$executeService$0(CompletionExecutor.java:171)
        at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: java.lang.ExceptionInInitializerError
        at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
        at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1160)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.classinitialization.ClassInitializationSupport.ensureClassInitialized(ClassInitializationSupport.java:177)
        ... 28 more
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
        ... 31 more
Caused by: java.lang.SecurityException: class "org.foo.SomeBean_Bean"'s signer information does not match signer information of other classes in the same package
        at java.base/java.lang.ClassLoader.checkCerts(ClassLoader.java:1173)
        at java.base/java.lang.ClassLoader.preDefineClass(ClassLoader.java:917)
        at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1025)
        at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageClassLoader.defineClass(NativeImageClassLoader.java:490)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageClassLoader.findClassViaClassPath(NativeImageClassLoader.java:442)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageClassLoader.loadClass(NativeImageClassLoader.java:629)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
        at io.quarkus.arc.setup.Default_ComponentsProvider.addBeans1(Unknown Source)
        at io.quarkus.arc.setup.Default_ComponentsProvider.getComponents(Unknown Source)
        at io.quarkus.arc.impl.ArcContainerImpl.<init>(ArcContainerImpl.java:128)
        at io.quarkus.arc.Arc.initialize(Arc.java:38)
        at io.quarkus.arc.runtime.ArcRecorder.initContainer(ArcRecorder.java:49)
        at io.quarkus.deployment.steps.ArcProcessor$initializeContainer1770303700.deploy_0(Unknown Source)
        at io.quarkus.deployment.steps.ArcProcessor$initializeContainer1770303700.deploy(Unknown Source)
        ... 32 more
```

### How to Reproduce?

See https://github.com/sithmein/signature-issue-native-build and the instructions in the README.

### Output of `uname -a` or `ver`

Linux crest 6.13.11-200.fc41.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Apr 10 19:02:09 UTC 2025 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "21.0.6" 2025-01-21

### Mandrel or GraalVM version (if different from Java)

Mandrel-23.1.6.0-Final

### Quarkus version or git rev

3.15.4

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.9.8

### Additional information

One way to solve this could be signing the generated Jars before passing them to the native build. However this only works if the library is "yours" i.e. the signing key is the same. It doesn't work for 3rd party libraries with beans. In such cases the signature information from all libs could be removed prior to starting the native build.

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/JarUnsigner.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/JarUnsigner.java
@@ -1,0 +1,164 @@
+package io.quarkus.deployment.pkg;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Enumeration;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.jar.Attributes;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.jar.JarOutputStream;
+import java.util.jar.Manifest;
+
+import org.jboss.logging.Logger;
+
+/**
+ * JarUnsigner is used to remove the signature from a jar file.
+ */
+public final class JarUnsigner {
+
+    private static final Logger log = Logger.getLogger(JarUnsigner.class);
+
+    private JarUnsigner() {
+        // utility class
+    }
+
+    /**
+     * Unsigns a jar file by removing the signature entries.
+     * If the JAR is not signed, it will simply copy the original JAR to the target path.
+     *
+     * @param jarPath the path to the jar file to unsign
+     * @param targetPath the path to the target jar file
+     * @throws IOException if an I/O error occurs
+     */
+    public static void unsignJar(Path jarPath, Path targetPath) throws IOException {
+        try (JarFile in = new JarFile(jarPath.toFile(), false)) {
+            Manifest manifest = in.getManifest();
+            boolean signed;
+            if (manifest != null) {
+                Map<String, Attributes> entries = manifest.getEntries();
+                signed = !entries.isEmpty();
+                // Remove signature entries
+                entries.clear();
+            } else {
+                signed = false;
+                manifest = new Manifest();
+            }
+            if (!signed) {
+                in.close();
+                log.debugf("JAR %s is not signed, skipping unsigning", jarPath);
+                Files.copy(jarPath, targetPath, java.nio.file.StandardCopyOption.REPLACE_EXISTING);
+            } else {
+                log.debugf("JAR %s is signed, removing signature", jarPath);
+                // Reusing buffer for performance reasons
+                byte[] buffer = new byte[10000];
+                try (JarOutputStream out = new JarOutputStream(Files.newOutputStream(targetPath))) {
+                    JarEntry manifestEntry = new JarEntry(JarFile.MANIFEST_NAME);
+                    // Set manifest time to epoch to always make the same jar
+                    manifestEntry.setTime(0);
+                    out.putNextEntry(manifestEntry);
+                    manifest.write(out);
+                    out.closeEntry();
+                    Enumeration<JarEntry> entries = in.entries();
+                    while (entries.hasMoreElements()) {
+                        JarEntry entry = entries.nextElement();
+                        String entryName = entry.getName();
+                        if (!entryName.equals(JarFile.MANIFEST_NAME)
+                                && !entryName.equals("META-INF/INDEX.LIST")
+                                && !isSignatureFile(entryName)) {
+                            entry.setCompressedSize(-1);
+                            out.putNextEntry(entry);
+                            try (InputStream inStream = in.getInputStream(entry)) {
+                                int r;
+                                while ((r = inStream.read(buffer)) > 0) {
+                                    out.write(buffer, 0, r);
+                                }
+                            } finally {
+                                out.closeEntry();
+                            }
+                        } else {
+                            log.debugf("Removed %s from %s", entryName, jarPath);
+                        }
+                    }
+                }
+            }
+            // let's make sure we keep the original timestamp
+            Files.setLastModifiedTime(targetPath, Files.getLastModifiedTime(jarPath));
+        }
+    }
+
+    /**
+     * Unsigns a jar file by removing the signature entries.
+     *
+     * @param jarPath the path to the jar file to unsign
+     * @param targetPath the path to the target jar file
+     * @param includePredicate a predicate to determine which entries to include in the target jar
+     * @throws IOException if an I/O error occurs
+     */
+    public static void unsignJar(Path jarPath, Path targetPath, Predicate<String> includePredicate) throws IOException {
+        // Reusing buffer for performance reasons
+        byte[] buffer = new byte[10000];
+        try (JarFile in = new JarFile(jarPath.toFile(), false)) {
+            Manifest manifest = in.getManifest();
+            boolean signed;
+            if (manifest != null) {
+                Map<String, Attributes> entries = manifest.getEntries();
+                signed = !entries.isEmpty();
+                // Remove signature entries
+                entries.clear();
+            } else {
+                signed = false;
+                manifest = new Manifest();
+            }
+            if (signed) {
+                log.debugf("JAR %s is signed, removing signature", jarPath);
+            }
+            try (JarOutputStream out = new JarOutputStream(Files.newOutputStream(targetPath))) {
+                JarEntry manifestEntry = new JarEntry(JarFile.MANIFEST_NAME);
+                // Set manifest time to epoch to always make the same jar
+                manifestEntry.setTime(0);
+                out.putNextEntry(manifestEntry);
+                manifest.write(out);
+                out.closeEntry();
+                Enumeration<JarEntry> entries = in.entries();
+                while (entries.hasMoreElements()) {
+                    JarEntry entry = entries.nextElement();
+                    String entryName = entry.getName();
+                    if (includePredicate.test(entryName)
+                            && !entryName.equals(JarFile.MANIFEST_NAME)
+                            && !entryName.equals("META-INF/INDEX.LIST")
+                            && !isSignatureFile(entryName)) {
+                        entry.setCompressedSize(-1);
+                        out.putNextEntry(entry);
+                        try (InputStream inStream = in.getInputStream(entry)) {
+                            int r;
+                            while ((r = inStream.read(buffer)) > 0) {
+                                out.write(buffer, 0, r);
+                            }
+                        } finally {
+                            out.closeEntry();
+                        }
+                    } else {
+                        log.debugf("Removed %s from %s", entryName, jarPath);
+                    }
+                }
+            }
+            // let's make sure we keep the original timestamp
+            Files.setLastModifiedTime(targetPath, Files.getLastModifiedTime(jarPath));
+        }
+    }
+
+    private static boolean isSignatureFile(String entry) {
+        entry = entry.toUpperCase();
+        if (entry.startsWith("META-INF/") && entry.indexOf('/', "META-INF/".length()) == -1) {
+            return entry.endsWith(".SF")
+                    || entry.endsWith(".DSA")
+                    || entry.endsWith(".RSA")
+                    || entry.endsWith(".EC");
+        }
+        return false;
+    }
+}

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java
@@ -43,9 +43,6 @@ import java.util.function.BooleanSupplier;
 import java.util.function.Consumer;
 import java.util.function.Predicate;
 import java.util.jar.Attributes;
-import java.util.jar.JarEntry;
-import java.util.jar.JarFile;
-import java.util.jar.JarOutputStream;
 import java.util.jar.Manifest;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
@@ -68,6 +65,7 @@ import io.quarkus.deployment.builditem.MainClassBuildItem;
 import io.quarkus.deployment.builditem.QuarkusBuildCloseablesBuildItem;
 import io.quarkus.deployment.builditem.TransformedClassesBuildItem;
 import io.quarkus.deployment.configuration.ClassLoadingConfig;
+import io.quarkus.deployment.pkg.JarUnsigner;
 import io.quarkus.deployment.pkg.PackageConfig;
 import io.quarkus.deployment.pkg.builditem.ArtifactResultBuildItem;
 import io.quarkus.deployment.pkg.builditem.BuildSystemTargetBuildItem;
@@ -959,7 +957,7 @@ public class JarResultBuildStep {
                 } else {
                     // we copy jars for which we remove entries to the same directory
                     // which seems a bit odd to me
-                    filterJarFile(resolvedDep, targetPath, removedFromThisArchive);
+                    JarUnsigner.unsignJar(resolvedDep, targetPath, Predicate.not(removedFromThisArchive::contains));
 
                     var list = new ArrayList<>(removedFromThisArchive);
                     Collections.sort(list);
@@ -1135,7 +1133,6 @@ public class JarResultBuildStep {
             TransformedClassesBuildItem transformedClasses, Path libDir,
             StringBuilder classPath, Collection<ResolvedDependency> appDeps, Map<String, List<byte[]>> services,
             Predicate<String> ignoredEntriesPredicate, Set<ArtifactKey> removedDependencies) throws IOException {
-
         for (ResolvedDependency appDep : appDeps) {
 
             // Exclude files that are not jars (typically, we can have XML files here, see https://github.com/quarkusio/quarkus/issues/2852)
@@ -1150,7 +1147,8 @@ public class JarResultBuildStep {
                     if (transformedFromThisArchive == null || transformedFromThisArchive.isEmpty()) {
                         final String fileName = appDep.getGroupId() + "." + resolvedDep.getFileName();
                         final Path targetPath = libDir.resolve(fileName);
-                        Files.copy(resolvedDep, targetPath, StandardCopyOption.REPLACE_EXISTING);
+                        // Unsign the jar before copying it
+                        JarUnsigner.unsignJar(resolvedDep, targetPath);
                         classPath.append(" lib/").append(fileName);
                     } else {
                         //we have transformed classes, we need to handle them correctly
@@ -1158,7 +1156,7 @@ public class JarResultBuildStep {
                                 + resolvedDep.getFileName();
                         final Path targetPath = libDir.resolve(fileName);
                         classPath.append(" lib/").append(fileName);
-                        filterJarFile(resolvedDep, targetPath, transformedFromThisArchive);
+                        JarUnsigner.unsignJar(resolvedDep, targetPath, Predicate.not(transformedFromThisArchive::contains));
                     }
                 } else {
                     // This case can happen when we are building a jar from inside the Quarkus repository
@@ -1270,66 +1268,6 @@ public class JarResultBuildStep {
                 }
             }
         }
-    }
-
-    static void filterJarFile(Path resolvedDep, Path targetPath, Set<String> transformedFromThisArchive) {
-        try {
-            byte[] buffer = new byte[10000];
-            try (JarFile in = new JarFile(resolvedDep.toFile(), false)) {
-                Manifest manifest = in.getManifest();
-                if (manifest != null) {
-                    // Remove signature entries
-                    manifest.getEntries().clear();
-                } else {
-                    manifest = new Manifest();
-                }
-                try (JarOutputStream out = new JarOutputStream(Files.newOutputStream(targetPath))) {
-                    JarEntry manifestEntry = new JarEntry(JarFile.MANIFEST_NAME);
-                    // Set manifest time to epoch to always make the same jar
-                    manifestEntry.setTime(0);
-                    out.putNextEntry(manifestEntry);
-                    manifest.write(out);
-                    out.closeEntry();
-                    Enumeration<JarEntry> entries = in.entries();
-                    while (entries.hasMoreElements()) {
-                        JarEntry entry = entries.nextElement();
-                        String entryName = entry.getName();
-                        if (!transformedFromThisArchive.contains(entryName)
-                                && !entryName.equals(JarFile.MANIFEST_NAME)
-                                && !entryName.equals("META-INF/INDEX.LIST")
-                                && !isSignatureFile(entryName)) {
-                            entry.setCompressedSize(-1);
-                            out.putNextEntry(entry);
-                            try (InputStream inStream = in.getInputStream(entry)) {
-                                int r = 0;
-                                while ((r = inStream.read(buffer)) > 0) {
-                                    out.write(buffer, 0, r);
-                                }
-                            } finally {
-                                out.closeEntry();
-                            }
-                        } else {
-                            log.debugf("Removed %s from %s", entryName, resolvedDep);
-                        }
-                    }
-                }
-                // let's make sure we keep the original timestamp
-                Files.setLastModifiedTime(targetPath, Files.getLastModifiedTime(resolvedDep));
-            }
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
-        }
-    }
-
-    private static boolean isSignatureFile(String entry) {
-        entry = entry.toUpperCase();
-        if (entry.startsWith("META-INF/") && entry.indexOf('/', "META-INF/".length()) == -1) {
-            return entry.endsWith(".SF")
-                    || entry.endsWith(".DSA")
-                    || entry.endsWith(".RSA")
-                    || entry.endsWith(".EC");
-        }
-        return false;
     }
 
     /**

--- a/core/deployment/src/test/java/io/quarkus/deployment/pkg/JarUnsignerTest.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/pkg/JarUnsignerTest.java
@@ -1,4 +1,4 @@
-package io.quarkus.deployment.pkg.steps;
+package io.quarkus.deployment.pkg;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
@@ -16,7 +16,6 @@ import java.security.cert.Certificate;
 import java.security.cert.CertificateException;
 import java.util.Calendar;
 import java.util.Date;
-import java.util.Set;
 import java.util.jar.JarEntry;
 import java.util.jar.JarFile;
 import java.util.jar.Manifest;
@@ -41,9 +40,9 @@ import org.junit.jupiter.api.io.TempDir;
 import jdk.security.jarsigner.JarSigner;
 
 /**
- * Test for {@link JarResultBuildStep}
+ * Test for {@link JarUnsigner}
  */
-class JarResultBuildStepTest {
+class JarUnsignerTest {
 
     @Test
     void should_unsign_jar_when_filtered(@TempDir Path tempDir) throws Exception {
@@ -58,7 +57,7 @@ class JarResultBuildStepTest {
                 FileOutputStream out = new FileOutputStream(signedJarPath.toFile())) {
             signer.sign(in, out);
         }
-        JarResultBuildStep.filterJarFile(signedJarPath, unsignedJarToTestPath, Set.of("java/lang/Integer.class"));
+        JarUnsigner.unsignJar(signedJarPath, unsignedJarToTestPath, p -> !p.equals("java/lang/Integer.class"));
         try (JarFile jarFile = new JarFile(unsignedJarToTestPath.toFile())) {
             assertThat(jarFile.stream().map(JarEntry::getName)).doesNotContain("META-INF/ECLIPSE_.RSA", "META-INF/ECLIPSE_.SF");
             // Check that the manifest is still present
@@ -76,7 +75,7 @@ class JarResultBuildStepTest {
         Path initialJar = tempDir.resolve("initial.jar");
         Path filteredJar = tempDir.resolve("filtered.jar");
         archive.as(ZipExporter.class).exportTo(new File(initialJar.toUri()), true);
-        JarResultBuildStep.filterJarFile(initialJar, filteredJar, Set.of("java/lang/Integer.class"));
+        JarUnsigner.unsignJar(initialJar, filteredJar, p -> !p.equals("java/lang/Integer.class"));
         try (JarFile jarFile = new JarFile(filteredJar.toFile())) {
             assertThat(jarFile.stream())
                     .filteredOn(jarEntry -> jarEntry.getName().equals(JarFile.MANIFEST_NAME))

